test/e2e/hypernode/network_topology_test.go

```diff
@@ -0,0 +1,463 @@
+					HighestTierAllowed: ptr.To(1),
+				},
+				Tasks: []e2eutil.TaskSpec{
+					{
+						Name:        "task-1",
+						Img:         e2eutil.DefaultNginxImage,
+						Req:         e2eutil.CPU2Mem2,
+						Min:         2,
+						Rep:         2,
+						Tolerations: tolerations,
+					},
+				},
+			}
+			topologyJob := e2eutil.CreateJob(ctx, job)
+
+			defer func() {
+				By("Delete job")
+				Expect(ctx.Vcclient.BatchV1alpha1().Jobs(topologyJob.Namespace).Delete(context.TODO(), topologyJob.Name, metav1.DeleteOptions{})).NotTo(HaveOccurred())
+			}()
+
+			By("Wait for job running")
+			Expect(e2eutil.WaitJobReady(ctx, topologyJob)).NotTo(HaveOccurred())
+
+			By("Verify pods are scheduled to s1")
+			Expect(e2eutil.VerifyPodScheduling(ctx, topologyJob, []string{"kwok-node-2", "kwok-node-3"})).NotTo(HaveOccurred())
+		})
+
+		It("Case 1.2: Schedule to s2 when s1's resources are insufficient and s2's score is higher", func() {
+			By("Create job that fits in s1's resources")
+
+			podSpecs := []e2eutil.PodSpec{
+				// allocate to s1(kwok-node-2, kwok-node-3) to make s1 has insufficient resources
+				{Name: "pod-0", Node: "kwok-node-2", Req: e2eutil.CPU4Mem4, Tolerations: tolerations},
+				{Name: "pod-1", Node: "kwok-node-3", Req: e2eutil.CPU4Mem4, Tolerations: tolerations},
+				// allocate to s2(kwok-node-4, kwok-node-5) to make s2 has enough resources and higher score
+				{Name: "pod-2", Node: "kwok-node-4", Req: e2eutil.CPU1Mem1, Tolerations: tolerations},
+			}
+
+			pods := make([]*v1.Pod, len(podSpecs))
+			for i, podSpec := range podSpecs {
+				pods[i] = e2eutil.CreatePod(ctx, podSpec)
+			}
+
+			defer func() {
+				for _, pod := range pods {
+					Expect(ctx.Kubeclient.CoreV1().Pods(pod.Namespace).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{})).NotTo(HaveOccurred())
+				}
+			}()
+
+			By("Wait for all pods to be ready")
+			for _, pod := range pods {
+				Expect(e2eutil.WaitPodReady(ctx, pod)).NotTo(HaveOccurred())
+			}
+
+			job := &e2eutil.JobSpec{
+				Name: "job-1",
+				NetworkTopology: &batchv1alpha1.NetworkTopologySpec{
+					Mode:               batchv1alpha1.HardNetworkTopologyMode,
+					HighestTierAllowed: ptr.To(1),
+				},
+				Tasks: []e2eutil.TaskSpec{
+					{
+						Name:        "task-1",
+						Img:         e2eutil.DefaultNginxImage,
+						Req:         e2eutil.CPU2Mem2,
+						Min:         2,
+						Rep:         2,
+						Tolerations: tolerations,
+					},
+				},
+			}
+
+			topologyJob := e2eutil.CreateJob(ctx, job)
+
+			defer func() {
+				By("Delete job")
+				Expect(ctx.Vcclient.BatchV1alpha1().Jobs(topologyJob.Namespace).Delete(context.TODO(), topologyJob.Name, metav1.DeleteOptions{})).NotTo(HaveOccurred())
+			}()
+
+			By("Wait for job running")
+			Expect(e2eutil.WaitJobReady(ctx, topologyJob)).NotTo(HaveOccurred())
+
+			By("Verify pods are scheduled to s2")
+			Expect(e2eutil.VerifyPodScheduling(ctx, topologyJob, []string{"kwok-node-4", "kwok-node-5"})).NotTo(HaveOccurred())
```

Comment:
I am a little curious here. Each kwok-node is 8U8G by default. kwok-node-2 and kwok-node-3 are currently allocated 4U4G each (pod0 and pod1). Why can't the vcjob of 2u2g for scheduling 2 replicas still be scheduled to kwok-node-2/kwok-node-3?

Reply:
I'm wondering that do we only need to specify `[]string{"kwok-node-4"}`here? Because we have also enabled binpack plugin